[KOGITO-246] Integrate Kogito Operator with Strimzi Operator

xiezhang7 opened the pull request:
Jira: https://issues.jboss.org/browse/KOGITO-246

Description: add options to deploy Strimzi Kafka Topic CRs in the Kogito Operator

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

radtriste commented on README.md:
```diff
@@ -257,71 +254,22 @@ my-cluster-zookeeper-nodes    ClusterIP   None             <none>        2181/TC
 #### Retrieve Kafka internal url
 
-Having configured the Kafka Operator, you'll need the Kafka service URI to install the Data Index Service. 
+Having configured the Kafka Operator, you'll need the Kafka service URI or custom resource name to install the Data Index Service. 
```
I would put 'Kafka Custom Resource name' to be more explicit

xiezhang7 replied:
Sure. Thanks.

radtriste commented on README.md:
```diff
@@ -333,6 +281,16 @@ my-cluster-kafka-bootstrap    ClusterIP   172.30.228.90    <none>        9091/TC
 
 In this case the Kafka Cluster service is `my-cluster-kafka-bootstrap:9092`.
 
+Or run the following command to take the custom resource name:
```
Also here :

"Or run the following command to retrieve the Kafka custom resource name:"
or
"Or run the following command to retrieve the Kafka instance name:"

radtriste commented on README.md:
```diff
@@ -365,7 +329,11 @@ spec:
   kafka:
     # the service name and port for the kafka cluster. Example: my-kafka-cluster:9092
-    serviceURI: my-cluster-kafka-bootstrap:9092
+    externalURI: my-cluster-kafka-bootstrap:9092
```
Why did we change to externalURI ?
Infinispan is still serviceURI ... We need some coherence here ;)

ricardozanini replied:
I'll change the Infinispan `Spec` on this JIRA: https://issues.jboss.org/browse/KOGITO-165
I told him to use `ExternalURI` since makes more sense now that we will use `service` for internal infra.

radtriste replied:
ok good for me then ;)

ricardozanini replied:
Actually I'll open another JIRA for this

radtriste commented on README.md:
```diff
@@ -235,13 +232,13 @@ Having the Infinispan Operator deployed, just jump to [Install Data Index Servic
 ![Strimzi Installed](docs/img/strimzi_installed.png?raw=true)
 
-Next, you need to create a Kafka cluster and a Kafka Topic for the Data Index Service to connect. Click on the name of the Strimzi Operator, then on `Kafka` tab and `Create Kafka`. Accept the default options to create a 3 node Kafka cluster. If it's a development environment, consider setting the Zookeeper and Kafka replicas to 1 to save resources.
+Next, you need to create a Kafka cluster for the Data Index Service to connect. Click on the name of the Strimzi Operator, then on `Kafka` tab and `Create Kafka`. Accept the default options to create a 3 node Kafka cluster. If it's a development environment, consider setting the Zookeeper and Kafka replicas to 1 to save resources.
```
Couldn't we do it via Kogito Operator ? Or is there any blocker ?
I mean we are already creating the Kafka Topics automatically and we are able to find the instance automatically if no provided (see cmd/kogito/command/install/data-index.go)

ricardozanini replied:
This should be handled by the Kogito Operator like is handled with Infinispan. A Kafka cluster should be created like Infinispan are with `KogitoInfra` controller.

xiezhang7 replied:
We will do this in https://issues.jboss.org/browse/KOGITO-614 .

radtriste commented on pkg/controller/kogitodataindex/resource/create_resources.go:
```diff
@@ -84,7 +88,8 @@ func CreateOrFetchResources(instance *v1alpha1.KogitoDataIndex, context resource
 		build(createStatefulSet).
 		build(createService).
-		buildOnOpenshift(createRoute)
+		buildOnOpenshift(createRoute).
```
What is this '.' EOL ?

xiezhang7 replied:
`buildOnOpenshift(createRoute)` returns `*kogitoDataIndexResourcesFactory`. We want to do `*kogitoDataIndexResourcesFactory.build(createKafkaTopic)` after `buildOnOpenshift(createRoute)`.  Here, `.` is used to [chain the calls](https://en.wikipedia.org/wiki/Method_chaining).

radtriste commented on cmd/kogito/command/install/data_index.go:
```diff
@@ -111,6 +111,11 @@ func (i *installDataIndexCommand) RegisterHook() {
+				log.Infof("kafka-url informed. Kafka will NOT be provisioned for you. Make sure that %s url is accessible from the cluster", i.flags.kafka.ExternalURI)
+			} else if len(i.flags.kafka.Instance) > 0 {
+				log.Infof("kafka-instance informed. Kafka will NOT be provisioned for you. Make sure Kafka custom resource %s is properly deployed in the project", i.flags.kafka.Instance)
+			}
```
In case no Kafka information delivered, I would add the information:

"No Kafka information has been given. It will be searched into the namespace"

ricardozanini commented on README.md:
```diff
@@ -235,13 +232,13 @@ Having the Infinispan Operator deployed, just jump to [Install Data Index Servic
 ### Deploy Strimzi
 
-Deploying [Strimzi](https://strimzi.io/) is easy since it's an Operator and should be available in the [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator). On OpenShift Web Console, go to the left menu, Catalog, OperatorHub and search for `Strimzi`.
+Find the [Strimzi](https://strimzi.io/) Operator in the [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator). On OpenShift Web Console, go to the left menu, `Operators`, `Installed Operators` and search for `Strimzi`.
```
If the OperatorHub is installed, the Kogito Operator will be installed with Strimzi, users shouldn't have to deploy it manually. The only condition to deploy Strimzi manually is on environments that OperatorHub is not available.

Plus, this README should be reviewed after this PR: https://github.com/kiegroup/kogito-cloud-operator/pull/109. That's about to be merged.

ricardozanini replied:
The "Deploy Strimzi" section should be moved to another section where Kogito Operator is deployed manually. I'd say to have a section named "Deploy Infinispan" with subsections like "Manually" and "via OperatorHub". The same for Strimzi. @sterobin wdyt?

radtriste replied:
Or we have a section in Readme for installation on OpenShift with no OperatorHub (aka mostly OC3) ?

xiezhang7 replied:
Since installing the Kogito operator via OperatorHub also installs Strimzi operator and Infinispan operator, this section actually only describes a simple example of deploying the **custom resources** of Strimzi. I think the details of manually deploying Strimzi and Infinispan operators could be found in their documentation. We could add some links to the documentation in the `Deploying to OpenShift 3.11` section.

radtriste commented:
I did some quick tests on that commit to check the operator.
Find below my remarks / questions.

*Installation with OLM*
- As Kafka Operator, it installs AMQ Streams instead of Strimzi. Is that normal ? (I would guess that anyway kafka will be present so it is normal, but just asking on desired behavior)
- As Infinispan Operator, I got first once DataGrid and second time Infinispan. same as Kafka, is that expected behavior ?
*Then running kogito install data-index*
- As expected, there is for now no direct instantiation of Kafka if it does not exists. We should have it right ?
- I get an info message saying Infinispan will be deployed via Operator. But it could as well guessed from existing instance or ? Or put a more explicit message saying that Infinispan will be guessed / deployed dependning on current configuration in namespace ?
- I get  will be instantiated but no error message saying that no Kafka exists ... Could you put one ? I had to check in logs to see this error:
`{"level":"error","ts":1574438492.314593,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kogitodataindex-controller","request":"tradisso-kogito-246/kogito-infra","error":"there's no Kafka instance URI found in the namespace and Kafka external URI is not specified, cannot deploy Data Index service","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tkogito-cloud-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:192\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:171\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}`
- If Kafka exist, I would expect a message saying that Kafka topics will be created if they don't exist already ?
- After that, checking in logs, I got the error below. What is the problem here ?
`{"level":"error","ts":1574437918.1387694,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kogitodataindex-controller","request":"tradisso-kogito-246/kogito-data-index","error":"the server could not find the requested resource (put kogitodataindices.app.kiegroup.org kogito-data-index)","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tkogito-cloud-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:192\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tkogito-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:171\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\tkogito-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}`
and this is my kogitodataindex:
`Name:         kogito-data-index
Namespace:    tradisso-kogito-246
Labels:       <none>
Annotations:  <none>
API Version:  app.kiegroup.org/v1alpha1
Kind:         KogitoDataIndex
Metadata:
  Creation Timestamp:  2019-11-22T15:59:42Z
  Generation:          2
  Resource Version:    1975314
  Self Link:           /apis/app.kiegroup.org/v1alpha1/namespaces/tradisso-kogito-246/kogitodataindices/kogito-data-index
  UID:                 1865187c-0d41-11ea-9a30-fa163efed44b
Spec:
  Image:  quay.io/kiegroup/kogito-data-index:latest
  Infinispan:
    Credentials:
      Password Key:    password
      Secret Name:     kogito-infinispan-credential
      Username Key:    username
    Service URI:       kogito-infinispan
    Use Kogito Infra:  true
  Kafka:
  Replicas:  1
Status:
  Deployment Status:
    Replicas:  0
  Service Status:
    Load Balancer:
Events:  <none>
`

*Tests without OLM*
Operator installation is ok and after installing both Infinispan and Kafka Operators, I got the same behavior.

I hope this helps :)
Did not want to put one comment per remark. But if needed, I can do it ;)

ricardozanini commented:
@radtriste great review! I hope @xiezhang7 find some time to work on it. Regarding these:


> - As Kafka Operator, it installs AMQ Streams instead of Strimzi. Is that normal ? (I would guess that anyway kafka will be present so it is normal, but just asking on desired behavior)
> - As Infinispan Operator, I got first once DataGrid and second time Infinispan. same as Kafka, is that expected behavior ?

It's the expected behavior. Unfortunately both teams set their CRDs with same Kind, Group and Version for downstream and upstream projects. So there's nothing we can do about it, maybe open an issue in theirs projects asking about this. I would guess that we are the first one trying to add them as dependencies, so we are facing this issue.

ricardozanini commented:
@radtriste also @xiezhang7 needs to rebase his commit since there are some changes that might affect the behavior of this feature.

radtriste commented:
> @radtriste also @xiezhang7 needs to rebase his commit since there are some changes that might affect the behavior of this feature.

np. Now that I have my scenario, it will be easy to check ;)

xiezhang7 commented on README.md:
```diff
@@ -218,13 +218,10 @@ For more information, see [Native X JVM Builds](#native-x-jvm-builds).
 The Kogito Operator is able to deploy the [Data Index Service](https://github.com/kiegroup/kogito-runtimes/wiki/Data-Index-Service) as a [Custom Resource](deploy/crds/app.kiegroup.org_v1alpha1_kogitodataindex_cr.yaml) (`KogitoDataIndex`). 
 
-Since Data Index Service depends on Kafka, it's necessary to manually deploy an Apache Kafka Cluster (we use [Strimzi](https://strimzi.io/)) in the same namespace.
+Data Index Service depends on Kafka, since 0.6.0 version, Kogito Operator will deploy an Apache Kafka Cluster (based on [Strimzi](https://strimzi.io/) operator) in the same namespace.
```
It only deploys the Strimzi operator and Kafka Topics. We will support deploying Kafka/cluster in https://issues.jboss.org/browse/KOGITO-614 .

xiezhang7 commented:
> Please find my comments below.
> 
> One other thing: I would change "Kafka Custom Resource" to "Kafka instance". Wdyt ?

@radtriste I have changed `Kafka Custom Resource` to `Kafka instance`. Thanks.

xiezhang7 commented:
> * As expected, there is for now no direct instantiation of Kafka if it does not exists. We should have it right ?

We will support this in https://issues.jboss.org/browse/KOGITO-614 .

xiezhang7 commented:
> * I get an info message saying Infinispan will be deployed via Operator. But it could as well guessed from existing instance or ? Or put a more explicit message saying that Infinispan will be guessed / deployed dependning on current configuration in namespace ?

I am not sure the design of Infinispan deployment. Maybe, @ricardozanini can clarify.

xiezhang7 commented:
> * After that, checking in logs, I got the error below. What is the problem here ?
>   `{"level":"error","ts":1574437918.1387694,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kogitodataindex-controller","request":"tradisso-kogito-246/kogito-data-index","error":"the server could not find the requested resource (put kogitodataindices.app.kiegroup.org kogito-data-index)"`

I have fixed this error.

xiezhang7 commented:
> * I get  will be instantiated but no error message saying that no Kafka exists ... Could you put one ? I had to check in logs to see this error:
>   `{"level":"error","ts":1574438492.314593,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kogitodataindex-controller","request":"tradisso-kogito-246/kogito-infra","error":"there's no Kafka instance URI found in the namespace and Kafka external URI is not specified, cannot deploy Data Index service"
> * If Kafka exist, I would expect a message saying that Kafka topics will be created if they don't exist already ?

Do you mean the `kogito` command ? The execution of `kogitio` command is asynchronous. It cannot return these messages.
We can only put the error messages in the `Status` of the KogitoDataIndex object.

ricardozanini commented:
> > * I get an info message saying Infinispan will be deployed via Operator. But it could as well guessed from existing instance or ? Or put a more explicit message saying that Infinispan will be guessed / deployed dependning on current configuration in namespace ?
> 
> I am not sure the design of Infinispan deployment. Maybe, @ricardozanini can clarify.

@radtriste Is this happening in https://github.com/kiegroup/kogito-cloud-operator/pull/116? Could you clarify?

ricardozanini commented on deploy/examples/data-index.yaml:
```diff
@@ -19,7 +19,7 @@ spec:
   kafka:
     # the service name and port for the kafka cluster. Example: my-kafka-cluster:9092
-    serviceURI: kogito-kafka-kafka-bootstrap:9092
+    externalURI: kogito-kafka-kafka-bootstrap:9092
```
We might stick with  `instance: kogito-kafka` instead.

xiezhang7 replied:
done

ricardozanini commented on pkg/apis/addtoscheme_app_v1alpha1.go:
```diff
@@ -30,7 +31,9 @@ import (
-	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme,
+	AddToSchemes = append(AddToSchemes,
+		v1alpha1.SchemeBuilder.AddToScheme,
+		v1beta1.SchemeBuilder.AddToScheme,
```
Could you add a more verbose name like `kafkabetav1` please?

ricardozanini commented on pkg/apis/kafka/v1beta1/kafka_types.go:
```diff
@@ -0,0 +1,68 @@
+
+// Kafka is the Schema for the kafkas API
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Kafka struct {
```
Since their operator is written in Java we don't have the go code, correct? Is this file generated based on their CRD or did you create it manually?

xiezhang7 replied:
I created it manually based on their CRD and the APIs written in Java,  and only kept the properties that kogito operator needs.

ricardozanini replied:
Do you think it’s feasible and or possible to automate this task? E.g. generate operator code based on others CRDs? I’m not asking to do this right now, but as an improvement for future versions.

xiezhang7 replied:
I think it is feasible.

ricardozanini commented on pkg/client/meta/meta.go:
```diff
@@ -111,6 +112,7 @@ func GetRegisteredSchema() *runtime.Scheme {
 		&v1alpha1.KogitoApp{}, &v1alpha1.KogitoAppList{},
 		&v1alpha1.KogitoDataIndex{}, &v1alpha1.KogitoDataIndexList{},
 		&v1alpha1.KogitoInfra{}, &v1alpha1.KogitoInfraList{})
+	s.AddKnownTypes(v1beta1.SchemeGroupVersion, &v1beta1.Kafka{}, &v1beta1.KafkaList{}, &v1beta1.KafkaTopic{}, &v1beta1.KafkaTopicList{})
```
same here, please add `kafkav1beta1` since we have a lot of imports from other CRDs.

xiezhang7 replied:
done

ricardozanini commented on pkg/controller/kogitodataindex/resource/kafka.go:
```diff
@@ -14,19 +14,153 @@
+		if len(kafka.Status.Listeners) > 0 {
+			for _, listenerStatus := range kafka.Status.Listeners {
+				if listenerStatus.Type == "plain" {
+					if len(listenerStatus.Addresses) > 0 {
```
Won't be more readable if we change to `listenerStatus.Type == "plain" && len(listenerStatus.Addresses) > 0` instead?

xiezhang7 replied:
done

ricardozanini commented on pkg/controller/kogitodataindex/resource/kafka.go:
```diff
@@ -14,19 +14,153 @@
+}
+
+func getKafkaServerReplicas(kafkaProp v1alpha1.KafkaConnectionProperties, namespace string, client *client.Client) (string, int32, error) {
+	if len(kafkaProp.ExternalURI) <= 0 && client.HasServerGroup("kafka.strimzi.io") {
```
Instead of calling `HasServerGroup`, why not a function `isKafkaAvailable`? Then we won't need to change the group name every where if this changes.

xiezhang7 replied:
done

ricardozanini commented on pkg/controller/kogitodataindex/resource/kafka.go:
```diff
@@ -14,19 +14,153 @@
+			Partitions: 10,
+			Config: map[string]string{
+				"retention.ms":  "604800000",
+				"segment.bytes": "1073741824",
```
Could you add those into constants?

xiezhang7 replied:
done

ricardozanini commented on pkg/controller/kogitodataindex/resource/manage_resources.go:
```diff
@@ -17,6 +17,7 @@ package resource
 import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1"
+	"github.com/kiegroup/kogito-cloud-operator/pkg/apis/kafka/v1beta1"
```
Could you please name this import to `kafkabetav1`?

ricardozanini commented:
On wait for @radtriste to verify and @spolti to approve.

radtriste commented:
@xiezhang7 Can you please rebase once again to incorporate Infinispan from @ricardozanini ?
There is also a small fix in it.

radtriste commented:
> > * I get  will be instantiated but no error message saying that no Kafka exists ... Could you put one ? I had to check in logs to see this error:
> >   `{"level":"error","ts":1574438492.314593,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kogitodataindex-controller","request":"tradisso-kogito-246/kogito-infra","error":"there's no Kafka instance URI found in the namespace and Kafka external URI is not specified, cannot deploy Data Index service"
> > * If Kafka exist, I would expect a message saying that Kafka topics will be created if they don't exist already ?
> 
> @radtriste do you mean the `kogito` command ? The execution of `kogitio` command is asynchronous. It cannot return these messages.
> We can only put the error messages in the `Status` of the KogitoDataIndex object.

We should just put a message that Kafka Topics will be created if they don't exist. So the user knows about that.

radtriste commented:
> @xiezhang7 Can you please rebase once again to incorporate Infinispan from @ricardozanini ?
> There is also a small fix in it.

In-between I gave it a try and it seems ok now.
(fixed locally the error that is corrected by https://github.com/kiegroup/kogito-cloud-operator/pull/116)
Rebase to be done, I will run the checks (does not take long) and we will be good to go ;)

xiezhang7 commented:
>  We should just put a message that Kafka Topics will be created if they don't exist. So the user knows about that.

@radtriste I have rebased the commit and updated the [messages](https://github.com/kiegroup/kogito-cloud-operator/blob/89c6b85e3f6e2954c31e42b8632e266839444947/cmd/kogito/command/install/data_index.go#L119) for the `kogito` command. Thanks.

ricardozanini commented:
@spolti could you please take a look also? Thanks!

ricardozanini commented:
Hi! I should merge https://github.com/kiegroup/kogito-cloud-operator/pull/117 this morning so @xiezhang7 can do a quick rebase before we merge this one. 

Thank you all!

ricardozanini commented:
@xiezhang7 once you rebase this and @spolti review, we're ready to go.

spolti commented on README.md:
```diff
@@ -193,13 +189,15 @@ After you deploy the Infinispan Operator, see [Deploying Strimzi](#deploying-str
-[Strimzi](https://strimzi.io/) is an Operator and should be available in the [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator). In the OpenShift Web Console, go to **Catalog** -> **OperatorHub** in the left menu and search for `Strimzi`.
+#### Kafka
+
+[Strimzi](https://strimzi.io/) is an Operator and should be available in the [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator). In the OpenShift Web Console, go to **Catalog** -> **Installed Operators** in the left menu and search for `Strimzi`.
```
Strimzi is not a operator but there is a Strimzi operator.
Strimzi is a kind of Kafka as a Service

ricardozanini replied:
@xiezhang7 maybe:
```markdown
[Strimzi](https://strimzi.io/) Operator should be available in the [OperatorHub](https://operatorhub.io/operator/strimzi-kafka-operator).
```

?

spolti replied:
+1

xiezhang7 replied:
@spolti @ricardozanini the [README](https://github.com/kiegroup/kogito-cloud-operator/blob/70c7f8d13b1ed70cbd5f15b0bb21745e036a7958/README.md#kafka) is updated. Thanks.

ricardozanini commented on pkg/controller/kogitodataindex/resource/kafka.go:
```diff
@@ -14,19 +14,160 @@
+
+	kafkaTopicConfigRetentionKey   string = "retention.ms"
+	kafkaTopicConfigSegmentKey     string = "segment.bytes"
+	kafkaTopicConfigRetentionValue string = "604800000"
```
They are the default values for topic's creation. Since we don't handle this to users to set upon creation, we're using defaults.

Users could update/edit this later via Strimzi.

ricardozanini commented on pkg/controller/kogitodataindex/resource/kafka.go:
```diff
@@ -14,19 +14,160 @@
+	kafkaTopicConfigRetentionKey   string = "retention.ms"
+	kafkaTopicConfigSegmentKey     string = "segment.bytes"
+	kafkaTopicConfigRetentionValue string = "604800000"
+	kafkaTopicConfigSegmentValue   string = "1073741824"
```
Replied

ricardozanini commented:
@radtriste you mentioned you will run some tests, right?

radtriste commented:
I am on it

radtriste commented:
When installing data-index, I have username and password correctly coming from `kogito-infinispan-credential - Secret` in env variables.
But I have `QUARKUS_INFINISPAN_CLIENT_SERVER_LIST=kogito-infinispan`. I think the port is missing here ...

radtriste commented:
This is my kogitodataindices:
```
apiVersion: app.kiegroup.org/v1alpha1
kind: KogitoDataIndex
metadata:
  creationTimestamp: "2019-12-03T16:39:02Z"
  generation: 5
  name: kogito-data-index
  namespace: kogito-246
  resourceVersion: "609845"
  selfLink: /apis/app.kiegroup.org/v1alpha1/namespaces/kogito-246/kogitodataindices/kogito-data-index
  uid: 69d30cfc-15eb-11ea-9a2d-52fdfc072182
spec:
  image: quay.io/tradisso/kogito-data-index:latest
  infinispan:
    credentials:
      passwordKey: password
      secretName: kogito-infinispan-credential
      usernameKey: username
    serviceURI: kogito-infinispan
    useKogitoInfra: true
  kafka: {}
  replicas: 1
status:
  conditions:
  - condition: OK
    lastTransitionTime: "2019-12-03T16:39:14Z"
    message: Deployment Finished
  - condition: Provisioning
    lastTransitionTime: "2019-12-03T16:39:14Z"
    message: Deployment Not Started
  - condition: OK
    lastTransitionTime: "2019-12-03T16:39:29Z"
    message: Deployment Finished
  dependenciesStatus:
  - OK
  deploymentStatus:
    collisionCount: 0
    currentReplicas: 1
    currentRevision: kogito-data-index-558c679bbb
    observedGeneration: 1
    readyReplicas: 1
    replicas: 1
    updateRevision: kogito-data-index-558c679bbb
    updatedReplicas: 1
  route: http://kogito-data-index-kogito-246.apps-crc.testing
  serviceStatus:
    loadBalancer: {}
```

and my kogitoinfra:
```
apiVersion: app.kiegroup.org/v1alpha1
kind: KogitoInfra
metadata:
  creationTimestamp: "2019-12-03T16:39:02Z"
  generation: 5
  name: kogito-infra
  namespace: kogito-246
  resourceVersion: "609707"
  selfLink: /apis/app.kiegroup.org/v1alpha1/namespaces/kogito-246/kogitoinfras/kogito-infra
  uid: 69d40794-15eb-11ea-9a2d-52fdfc072182
spec:
  installInfinispan: true
status:
  condition:
    lastTransitionTime: "2019-12-03T16:39:03Z"
    status: "True"
    type: Success
  infinispan:
    condition:
    - lastTransitionTime: "2019-12-03T16:39:03Z"
      status: "True"
      type: Success
    credentialSecret: kogito-infinispan-credential
    name: kogito-infinispan
    service: kogito-infinispan
```

ricardozanini commented:
This shouldn’t be a problem since the client is smart enough to figure out the default port. That’s why there’s no error. 

I will add the port in the next PR.

radtriste commented:
Ok.
No error in data-index logs indeed.

Note that this does not happen for a `kogitoapp` which requires infinispan... I get `QUARKUS_INFINISPAN_CLIENT_SERVER_LIST=kogito-infinispan:11222` in env variables.

ricardozanini commented:
@radtriste:
https://github.com/kiegroup/kogito-cloud-operator/blob/master/pkg/controller/kogitodataindex/kogitodataindex_controller.go#L217
